### ISSUE TITLE
Implement `MerkleStore` data structure

### ISSUE THREAD

**Issue by bobbinth** [0xPolygonMiden/crypto#89]
Update by @grjte on 2023/03/09:
we now intend to implement `MerkleStore` to hold all Merkle-based data in the advice provider as described by @bobbinth in this comment below: https://github.com/0xPolygonMiden/crypto/issues/89#issuecomment-1460756669

# Original issue description

One of the Merkle tree-based data structures we have is a `MerklePathSet`. This data structure can be viewed as a set of Merkle paths resolving to the same root. Current limitation is that all paths must be of the same length - but I'm wondering if we can remove this limitation.

### Partial Merkle tree
If we do, this updated data structure could be named `PartialMerleTree` and it could look something like this:

![image](https://user-images.githubusercontent.com/43513081/223598416-633387fe-0bde-421a-981b-d035859a755d.png)

In the above, we know node value at `(index = 0, depth = 2)` but we don't know node value at `(index = 0, depth = 3)`.

I think such a data structure shouldn't be too difficult to implement, and the nice thing is that it will work naturally with our `mtree_get` and `mtree_set` instructions for any node depth.

For `mtree_set` specifically, we can easily replace a node in one `PartialMerkleTree` with a root of another `PartialMerkleTree` (assuming the combined depth doesn't exceed 64). However, we would not be able to replace a node in a `PartialMerkleTree` with a root of a Sparse Merkle tree (or rather that would be a very expensive operation).

So, I'm wondering if we could take this a step further.

### Merkle Store
A `MerkleStore` could be data structure similar to `PartialMerkleTreee` but such that we allow nodes to be roots of empty subtrees at no extra cost. For example:

![image](https://user-images.githubusercontent.com/43513081/223610201-9767fc3e-949e-4910-95d2-5239e364640a.png)

In the above, `E2` is a root of an empty subtree of depth 2. Instead of storing all the leaves of this subtree explicitly, we just store `E2`.

If we can implement this data structure efficiently (and I don't know if we'll run into some complications here), we could potentially replace other Merkle tree-based data structures with it. Specifically:

- `SimpleSmt` would be a special case of this data structure such that we know values of all leaves.
- `MerkleTree` would also be a special case  such that we know all the leaves and most of them are non-zero.
- `TieredSmt` would also be a special case such that we allow leaves only on specific tiers and impose additional structure on the leaf nodes.
- And as already mentioned, this data structure would directly replace `MerklePathSet`.

The main benefit (the way I see it) is that we end up with a single data structure which is easily "composable" - i.e., we can relatively easily replace a node in `MerkleStore` with a root of another `MerkleStore` to combine two trees.

**Comment by frisitano** [comment]
I think consolidating the data structures into a unified generic implementation (and specialising where necessary) seems reasonable given the high level of overlap between the data structures.  I think the implementation details here become very important and have significant implications in terms of efficiency and feature set.

## Features

I will start of by defining some high level features that could be useful to have in the `MerkleStore`:

- Versioned Merkle trees
- Efficient node storage
- Database recording and storage proofs
- Lazy loading
- Caching

### Versioned Merkle trees and efficient node storage
It may be useful to be able to query different versions of a merkle tree. However, we want to ensure that versioning is efficient. It would not be practical to copy all nodes of the tree for each version as this would result in a huge amount of data duplication which is very inefficient.

The first question we have to answer is how do we identify a particular version of a merkle tree and the easy answer is it's root.  Each unique tree will have a unique root.  

The next question is how do store a merkle tree in such a way that enables efficient versioning.  First of all lets consider why using node indexes as a node identifier will not work.  Lets say we have two different tree versions with two different roots and store the node data for each tree as a map of `node_index` -> (`left_child_hash`, `right_child_hash`).  The problem with this implementation is that we have no way of identifying which nodes are different in each of the tree maps because the keys `node_index` are not unique. This means that to store both versions we would need to keep a complete copy of all nodes.  This is extremely inefficient given the fact that there will likely be significant overlap between the two tree versions.

A solution to this problem is to modify the way in which we store the node data.  Instead of using the `node_index` as a key, we use the `node_hash` (where `node_hash = H(left_child_hash, right_child_hash)`) such that we have a map of `node_hash` -> (`left_child_hash`, `right_child_hash`).  Now the identifier of a node is unique based on it's contents.  We now have to consider how we can traverse the tree for different versions.  The way in which this is done is using a tuple of (`root`, `key`).  The `root` is used to lookup a node from the map, this will return a tuple of (`left_child_hash`, `right_child_hash`).  We then use the first bit of the key to choose which `child_hash` to use for the next map lookup.  We then choose the appropriate child and and use the hash to lookup the next node from the map. This is repeated until we reach a leaf.

Under this construction for each version we only store an addition `log(d)` nodes in the store as opposed to the required `2 * d` additional nodes required under the index based approach.  Nodes that are the same across different tree versions are only stored once.

### Database recording and storage proofs
It could be of value to record database reads, using an object we will refer to as a recorder, such that we can convert the recorder into a storage proof.  This would be particularly useful if we want to separate the responsibilities of execution and proving.  Lets say we want to execute a transaction (or produce a block) on server A but then prove it on server B (but server B doesn't have a copy of the database - it's stateless).  What we would need to do is every time we perform a lookup from the database backend we record the node in the recorder.  The recorder will hold all nodes we have retrieved from the DB during the execution of a transaction (or during block production).  Upon completion of the transaction we can then convert this recorder into a storage proof.  The storage proof will only need to store a byte vector of children hashes e.g. [`child_hash_1`, `child_hash_2`, `child_hash_3`, `child_hash_4`].  We can then reconstruct the database using this data - e.g. `key = H(child_hash_1, child_hash2)` / `value = (child_hash_1, child_hash2)` etc.  The prover would reconstruct an in memory db using this pattern and use it to seed the advice provider.

 ### Lazy loading and caching
It is not practical to know ahead of time exactly which data is going to be required for a particular transaction.  As such we need a mechanism for dynamically loading this data at runtime from the db.  One way in which this could be done is by providing a `DataStore` trait which implements the appropriate functions for fetching and inserting data.  We should have different implementations for the `DataStore` for db and in memory backends.  The way this would work is the `MerkleTreeStore` makes a request to the `DataStore` for a node with a particular key.  It is then the responsibility of the `DataStore` to return this.  For the db backed `DataStore` we would probably want to implement an in memory cache such that once we have looked up a node we do not have to do a second db lookup if we need to use it again. 

Most of the features I have described above have been implemented in this library I developed [here](https://github.com/frisitano/merkle-tree-db/tree/master).

**Comment by vlopes11** [comment]
The approach looks interesting, but I'm curious to see how efficient we can make it in order to produce merkle paths.

Given we have the following depth 4 tree

![image](https://user-images.githubusercontent.com/8730839/223706614-1d16cc7c-0a08-4602-8a46-9d8dc44c855a.png)

If we insert a leaf, we will produce the following path

![image](https://user-images.githubusercontent.com/8730839/223706938-4d047983-21f8-4ee6-a7a4-7344e175de41.png)

So far, its trivial. We will have a partial merkle tree with a single path. Then, we insert another leaf

![image](https://user-images.githubusercontent.com/8730839/223707171-6eaebcb9-ea4e-4a22-bcc5-44fc8e66c21a.png)

With this, we have green will have a path conflict with orange. This means they will need to have a shared path to the root, defined by purple. This isn't a hard problem as well given that when computing a full merkle path, if we end the current path at a given depth, then we just lookup the storage for the continuation of that path. Meaning: a path that starts at green will end at purple.

The Partial Merkle Trees will look like this

![image](https://user-images.githubusercontent.com/8730839/223707603-b6e12d1e-8047-402c-91d3-c7426fb0ac55.png)

The part that concerns me is the following. Suppose we want to produce a Merkle path from the light gray leaf that doesn't exist in the set

![image](https://user-images.githubusercontent.com/8730839/223707859-b9a14d3a-1462-4b3e-8dae-2976da8fb8f8.png)

The Merkle path will have to look like this

![image](https://user-images.githubusercontent.com/8730839/223707949-3cfe818d-dc76-455c-b026-56c0ccf4f7f7.png)

For convenience, I'll name `zero constant` the value of an empty sub-tree for a given depth.

As discussed above, it is trivial to produce paths for existing trees and the premature termination of a path implies we have another path that starts at that node. However, for empty subtrees, this assumption won't hold. The only solution I see now is to perform a full scan on the partial merkle tree sets until we find a node, falling back to the zero constant. We will have to repeat this over and over for every zero constant, until we finish the construction of the path, as they might randomly intersect with existing paths.

If we have to implement the way I described, it will be a very expensive operation for large sets.

**Comment by vlopes11** [comment]
> Can you elaborate on this a bit please? What do you mean by perform a full scan of the partial merkle tree sets?

Sure! To produce a Merkle Path from a leaf that doesn't exist, we need a query function that will, for each sibling, query the data struct for a node value for a given depth and index. Let's name the node in the example above

- `A`: The light gray node, indexed by (depth = 4, index = 12)
- `B`: The sibling of `A`, indexed by (depth = 4, index = 13)
- `C`: The merkle sibling of the next depth, indexed by (depth = 3, index = 7)
- `D`: One member of the purple path, indexed by (depth = 2, index = 2)
- `E`: The final element of the path, indexed by (depth = 1, index = 0)

To compute the first element of the path to `A`, we need to fetch the value of `B`. For that, we need to, first, check if `B` exists in any of the paths of the set.

The heart of the problem is: that computation is not trivial. We need to scan every path of the set and look for `B`. If we don't find any, then we fallback to `zero constant`. To achieve an efficient implementation, we need an efficient function to provide that link.

We proceed with the same logic for the remainder elements. When we get to `D`, it will happen to exist in one of the paths, so we just return it.

**Comment by frisitano** [comment]
> > Can you elaborate on this a bit please? What do you mean by perform a full scan of the partial merkle tree sets?
> 
> Sure! To produce a Merkle Path from a leaf that doesn't exist, we need a query function that will, for each sibling, query the data structure for a node value for a given depth and index. Let's name the nodes of the example above
> 
> * `A`: The light gray node, indexed by (depth = 4, index = 12)
> * `B`: The sibling of `A`, indexed by (depth = 4, index = 13)
> * `C`: The merkle sibling of the next depth, indexed by (depth = 3, index = 7)
> * `D`: One member of the purple path, indexed by (depth = 2, index = 2)
> * `E`: The final element of the path, indexed by (depth = 1, index = 0)
> 
> To compute the first element of the path to `A`, we need to fetch the value of `B`. For that, we need to, first, check if `B` exists in any of the paths of the set.
> 
> The heart of the problem is: that computation is not trivial. We need to scan every path of the set and look for `B`. If we don't find any, then we fallback to `zero constant`. To achieve an efficient implementation, we need an efficient function to provide that link.
> 
> We proceed with the same logic for the remainder elements `C`, `D`, and `E`. When we get to `D`, it will happen to exist in one of the paths, so we just return it.
> 
> It is important to keep in mind that a merkle proof for a non-existing leaf will never be a regular merkle path as we always take the sibling node. So we need to compute node by node.

Did you read my post above and review the library?  non-inclusion becomes trivial under this implementation.

**Comment by vlopes11** [comment]
@frisitano sorry, I don't fully grasp it as solution. A couple of preliminary questions:

- Did I understand correctly the proposal of this issue, that is to store sequences of nodes (i.e. paths), instead of individual nodes?
- If we are to store paths instead of nodes, it's not clear how we can efficiently index a node.

**Comment by frisitano** [comment]
> @frisitano sorry, I don't fully grasp it as solution. A couple of preliminary questions:
> 
> * Did I understand correctly the proposal of this issue, that is to store sequences of nodes (i.e. paths), instead of individual nodes?
> * If we are to store paths instead of nodes, it's not clear how we can efficiently index a node.

The misunderstanding is most likely mine based on my lack of experience working with the `MerklePathSet`.

**Comment by vlopes11** [comment]
@frisitano no worries! Let's wait for @bobbinth to clarify what of the two invariants he proposed

If we are to store paths, one efficient way is to have a mapping of the index of the end of the path to the path itself. This way, we start from the root, and we progress towards the desired node. It will be more expensive than a simple index -> path, but will be logarithmic

**Comment by frisitano** [comment]
> @frisitano no worries! Let's wait for @bobbinth to clarify what of the two invariants he proposed
> 
> If we are to store paths, one efficient way is to have a mapping of the index of the end of the path to the path itself. This way, we start from the root, and we progress towards the desired node. It will be more expensive than a simple index -> path, but will be logarithmic

Yes makes sense, after re-reading the proposal I think your interpretation is correct.

Personally I think it makes sense to start thinking about the features I described.  These features will become quite important as we move towards building the node and the associated infrastructure.  Right now the Merkle objects act as both storage structures and interfaces.  I think this should change and storage should be delegated to a storage backend which they read from and write to.  This should be implemented as some `DataStore` trait variant that the Merkle objects leverage.

**Comment by vlopes11** [comment]
@frisitano I agree entirely! I started a discussion [here](https://github.com/0xPolygonMiden/crypto/issues/30), and the link you provided seems to be a solution to that.

We might have a broader problem is we want a general storage as we might want to separate trees (otherwise I can maliciously insert a node into another tree since they share the same storage), but that would be a separate discussion.

**Comment by frisitano** [comment]
> @frisitano I agree entirely! I started a discussion [here](https://github.com/0xPolygonMiden/crypto/issues/30), and the link you provided seems to be a solution to that.

Awesome I will take a look at that! It's great to get that discussion going.

> We might have a broader problem is we want a general storage as we might want to separate trees (otherwise I can maliciously insert a node into another tree since they share the same storage), but that would be a separate discussion.

How would you go about maliciously inserting a node?  In any case I think it is good practice to partition the trees and a simple mechanism for this would be to add a prefix for to the key for each domain.  E.g. For an account storage all keys associated with the account are prefixed with the AccountId.

**Comment by bobbinth** [comment]
Based on offline discussions, I think we can implement `MerkleStore` to hold all Merkle-based data in the advice provider. It could look as follows:

```Rust
pub struct MerkleStore {
    nodes: BTreeMap<Digest, Node>
}

impl MerkleStore {

    // CONSTRUCTORS
    // --------------------------------------------------------------------------------------------

    pub fn add_merkle_tree(&mut self, leaves: Vec<Word>) -> Result<Word, MerkleError> {
        todo!()
    }

    pub fn add_sparse_merkle_tree<R, I>(&mut self, entries: R) -> Result<Word, MerkleError>
    where
        R: IntoIterator<IntoIter = I>,
        I: Iterator<Item = (u64, Word)> + ExactSizeIterator,
    {
        todo!()
    }

    pub fn add_merkle_paths<I>(&mut self, paths: I) -> Result<Word, MerkleError>
    where
        I: IntoIterator<Item = (u64, Word, MerklePath)>,
    {
        todo!()
    }

    // PUBLIC ACCESSORS
    // --------------------------------------------------------------------------------------------

    pub fn get_node(&self, root: Word, index: NodeIndex) -> Result<Word, MerkleError> {
        todo!()
    }

    pub fn get_path(&self, root: Word, index: NodeIndex) -> Result<MerklePath, MerkleError> {
        todo!()
    }

    // DATA MUTATORS
    // --------------------------------------------------------------------------------------------

    pub fn set_node(&mut self, root: Word, index: NodeIndex, value: Word) -> Result<(), MerkleError> {
        todo!()
    }

    pub fn merge_roots(&mut self, root1: Word, root2: Word) -> Result<Word, MerkleError> {
        todo!()
    }
}
```

Where `Digest` is `RpoDigest` and `Node` looks as follows:
```Rust
pub struct Node {
    left: Digest,
    right: Digest,
}
```

With this data structure we should be able to satisfy all the needs of the in-memory advice provider including MMR, SMT etc. (tough, I might have missed something).

**Comment by vlopes11** [comment]
@bobbinth can you specify how is the digest from `nodes: BTreeMap<Digest, Node>` obtained?

**Comment by bobbinth** [comment]
We just hash node contents - i.e., `merge(node.left, node.right) -> digest`.

**Comment by vlopes11** [comment]
@bobbinth I think this new interface is compatible with the approach recommended by @frisitano . Do you think we should go that direction? If we tackle the issue now to generalize the merkle objects, all the tree variants will be easier to implement (tsmt, mmr, etc) as they will be just specific algorithms

**Comment by vlopes11** [comment]
@bobbinth this will depend on how we interpret what a node is. If we say that a node is anything with a depth smaller than the max, then we will have internal nodes acting as leaves (i.e. depths 16, 32, 48) for the TSMT.

However, if we allow such storage to be indexed by depth and node, and we assume any index can hold such pre-image, then we can treat leaves and nodes as equals for that implementation, and it would work.

The advice provider implementation can be smart enough to fetch a node, understand if its internal or not - if internal, just return the node value of its parents; if leaf, return either the key/value pair that created that leaf, or the ordered list

**Comment by bobbinth** [comment]
I think within this data structure we define a node as something that has a left and a right child. Anything which doesn't have such a structure is a leaf and leaves can be located at any depth. Basically, in my mind, the advice provider separate parts for storing nodes (in the Merkle store) and all other data for which we might want to get pre-images (advice map).

**Comment by frisitano** [comment]
Is there a particular reason you are against storing leaves in the `MerkleStore`.  In my mind all nodes should be stored in the same structure (in this case the `BTreeMap`).  If we have different node variants then they should be represented using an enum, something like [this](https://github.com/frisitano/merkle-tree-db/blob/master/src/node.rs#L103-L116). In my mind normal leaf nodes and TSMT internal leaf nodes should be included in this enum and the values (pre-images) should be stored in the enums as well.  The `MerkleStore` is then responsible for implementing the algorithms for lookup and providing pre-images. That way all data is encapsulated in a single struct.  One of the downsides of this approach is we can not achieve constant time lookup for a particular node, you will always have to traverse to the node from the root.  However, I don't think this is a big problem as I would assume that we always have to provide authentication data when providing a pre-image anyway, so traversal / lookup of all nodes is a requirement in any case?

**Comment by bobbinth** [comment]
I'd like to keep the `MerkleStore` as un-opinionated as possible and also have interface between the VM and the advice provider relatively simple.  Different use cases can impose different restrictions on what the shape of a leaf could be, and ideally we'd be able to support them without modifying the VM code.

We structure `MerkleStore` in a way which supports arbitrary leaves but then we'd be duplicating the functionality of the advice map. Basically, we need to have the advice map anyway for many other use cases, might as well use it for leaf pre-images as well.

**Comment by bobbinth** [comment]
Closed by #93.
